shipstation_integration/orders.py

```diff
@@ -285,23 +292,57 @@ def create_erpnext_order(
 
 	so.save()
 
-	before_submit_hook = frappe.get_hooks("update_shipstation_order_before_submit")
-	if before_submit_hook:
-		so = frappe.get_attr(before_submit_hook[0])(store, so, order)
-		if so:
-			so.save()
-	if so:
-		so.submit()
-		frappe.db.commit()
-
-	after_submit_hook = frappe.get_hooks("update_shipstation_order_after_submit")
-	if before_submit_hook:
-		frappe.get_attr(after_submit_hook[0])(store, so, order)
-		frappe.db.commit()
+	set_status(so, order, store)
 
 	return so.name if so else None
 
 
+def set_status(
+	so: Union["SalesOrder", dict], order: "ShipStationOrder", store: "ShipstationStore"
+) -> None:
+
+	status_mapping = {
+		"awaiting_payment": ("Draft", 0),
+		"awaiting_shipment": ("To Deliver", 1),
+		"shipped": ("Completed", 1),
+		"on_hold": ("On Hold", 1),
+		"cancelled": ("Cancelled", 2),
+		"pending_fulfillment": ("To Deliver and Bill", 1),
+	}
+
+	new_status, new_docstatus = status_mapping.get(order.order_status, ("Draft", 0))
+
+	if so.status == new_status and so.docstatus == new_docstatus:
+		return
+
+	if isinstance(so, dict):
+		so = frappe.get_doc("Sales Order", so.get("name"))
+
+	so.status = new_status
+
+	if new_docstatus == 1 and so.docstatus == 0:
+		before_submit_hook = frappe.get_hooks("update_shipstation_order_before_submit")
+		if before_submit_hook:
+			so = frappe.get_attr(before_submit_hook[0])(store, so, order)
+			if so:
+				so.save()
+
+		so.submit()
+
+		after_submit_hook = frappe.get_hooks("update_shipstation_order_after_submit")
+		if after_submit_hook:
+			frappe.get_attr(after_submit_hook[0])(store, so, order)
+	elif new_docstatus == 2 and so.docstatus != 2:
+		so.cancel()
+	elif new_docstatus == 0 and so.docstatus != 0:
+		so.docstatus = 0
+		so.save()
```

Comment:
I think we should investigate in what cases would a document need to be un-submitted like this, since that's not conventional flow.

Reply:
Right... this might not be necessarily helpful. It seems more crucial to react to statuses beyond draft state, rather than forcefully reverting to draft.

Reply:
removed via 5f16468466fd8b943e934198f1d3a650ed3c75d8